File: ios/Runner/NativeShowcaseTelemetryStreamHandler.swift
import Flutter
import Foundation

final class NativeShowcaseTelemetryStreamHandler: NSObject, FlutterStreamHandler {
  private let workerQueue = DispatchQueue(label: "com.example.flutter_bloc_app.telemetry.worker")
  private var emitTimer: DispatchSourceTimer?
  private var sampleTimer: DispatchSourceTimer?
  private var eventSink: FlutterEventSink?

  private var sequence = 0
  private var sourceTick = 0
  private var windowSampleCount = 0
  private var windowSampleSum = 0.0
  private var windowDroppedCount = 0
  private var lastSampleValue = Double.nan

  private let sourceRateHz = 60
  private let deliveredRateHz = 4
  private let deliveryWindowMs = 250
  private let sampleEpsilon = 0.01

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    startWorkers()
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    stopWorkers()
    eventSink = nil
    resetAggregation()
    return nil
  }

  private func startWorkers() {
    let sampleInterval = DispatchTimeInterval.nanoseconds(Int(1_000_000_000 / sourceRateHz))
    let emitInterval = DispatchTimeInterval.milliseconds(deliveryWindowMs)

    let sampleTimer = DispatchSource.makeTimerSource(queue: workerQueue)
    sampleTimer.schedule(deadline: .now(), repeating: sampleInterval)
    sampleTimer.setEventHandler { [weak self] in
      self?.collectSample()
    }
    sampleTimer.resume()
    self.sampleTimer = sampleTimer

    let emitTimer = DispatchSource.makeTimerSource(queue: workerQueue)
    emitTimer.schedule(deadline: .now() + emitInterval, repeating: emitInterval)
    emitTimer.setEventHandler { [weak self] in
      self?.emitAggregate()
    }
    emitTimer.resume()
    self.emitTimer = emitTimer
  }

  private func stopWorkers() {
    sampleTimer?.cancel()
    sampleTimer = nil
    emitTimer?.cancel()
    emitTimer = nil
  }

  private func collectSample() {
    let sampleValue = demoSampleValue(sourceTick)
    sourceTick += 1

    if !lastSampleValue.isNaN && abs(sampleValue - lastSampleValue) < sampleEpsilon {
      windowDroppedCount += 1
    } else {
      windowSampleCount += 1
      windowSampleSum += sampleValue
      lastSampleValue = sampleValue
    }
  }

  private func emitAggregate() {
    guard windowSampleCount > 0 else {
      windowDroppedCount = 0
      return
    }

    sequence += 1
    let averageValue = windowSampleSum / Double(windowSampleCount)
    let payload: [String: Any] = [
      "sequence": sequence,
      "sampleCount": windowSampleCount,
      "averageValue": averageValue,
      "sourceRateHz": sourceRateHz,
      "deliveredRateHz": deliveredRateHz,
      "droppedCount": windowDroppedCount,
      "emittedAtMillis": Int(Date().timeIntervalSince1970 * 1000),
    ]

    DispatchQueue.main.async { [weak self] in
      self?.eventSink?(payload)
    }

    windowSampleCount = 0
    windowSampleSum = 0
    windowDroppedCount = 0
  }

  private func resetAggregation() {
    sequence = 0
    sourceTick = 0
    windowSampleCount = 0
    windowSampleSum = 0
    windowDroppedCount = 0
    lastSampleValue = .nan
  }

  private func demoSampleValue(_ tick: Int) -> Double {
    let wave = sin(Double(tick) * 0.15) * 50.0
    let counter = Double(tick % 10)
    return wave + counter
  }
}

File: android/app/src/main/kotlin/com/ilkersevim/blocflutter/NativeShowcaseTelemetryStreamHandler.kt
package com.ilkersevim.blocflutter

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import kotlin.math.sin

class NativeShowcaseTelemetryStreamHandler : EventChannel.StreamHandler {
  private var workerThread: HandlerThread? = null
  private var workerHandler: Handler? = null
  private var mainHandler: Handler? = null
  private var eventSink: EventChannel.EventSink? = null
  private var sampleRunnable: Runnable? = null
  private var emitRunnable: Runnable? = null
  private var sessionGeneration = 0L

  private var sequence = 0
  private var sourceTick = 0
  private var windowSampleCount = 0
  private var windowSampleSum = 0.0
  private var windowDroppedCount = 0
  private var lastSampleValue = Double.NaN

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    stopSession()
    sessionGeneration += 1
    val generation = sessionGeneration
    eventSink = events
    mainHandler = Handler(Looper.getMainLooper())
    workerThread = HandlerThread("NativeShowcaseTelemetry").also { it.start() }
    workerHandler = Handler(workerThread!!.looper)

    val sampleIntervalMs = 1000.0 / SOURCE_RATE_HZ
    val emitIntervalMs = DELIVERY_WINDOW_MS.toLong()

    sampleRunnable = object : Runnable {
      override fun run() {
        if (generation != sessionGeneration) {
          return
        }

        val sampleValue = demoSampleValue(sourceTick)
        sourceTick += 1

        if (!lastSampleValue.isNaN() &&
          kotlin.math.abs(sampleValue - lastSampleValue) < SAMPLE_EPSILON
        ) {
          windowDroppedCount += 1
        } else {
          windowSampleCount += 1
          windowSampleSum += sampleValue
          lastSampleValue = sampleValue
        }

        workerHandler?.postDelayed(this, sampleIntervalMs.toLong().coerceAtLeast(1L))
      }
    }

    emitRunnable = object : Runnable {
      override fun run() {
        if (generation != sessionGeneration) {
          return
        }

        if (windowSampleCount > 0) {
          sequence += 1
          val averageValue = windowSampleSum / windowSampleCount
          val payload = mapOf(
            "sequence" to sequence,
            "sampleCount" to windowSampleCount,
            "averageValue" to averageValue,
            "sourceRateHz" to SOURCE_RATE_HZ,
            "deliveredRateHz" to DELIVERED_RATE_HZ,
            "droppedCount" to windowDroppedCount,
            "emittedAtMillis" to System.currentTimeMillis(),
          )
          mainHandler?.post {
            if (generation == sessionGeneration) {
              eventSink?.success(payload)
            }
          }
        }

        windowSampleCount = 0
        windowSampleSum = 0.0
        windowDroppedCount = 0

        workerHandler?.postDelayed(this, emitIntervalMs)
      }
    }

    workerHandler?.post(sampleRunnable!!)
    workerHandler?.postDelayed(emitRunnable!!, emitIntervalMs)
  }

  override fun onCancel(arguments: Any?) {
    stopSession()
  }

  private fun stopSession() {
    sessionGeneration += 1
    sampleRunnable?.let { workerHandler?.removeCallbacks(it) }
    emitRunnable?.let { workerHandler?.removeCallbacks(it) }
    sampleRunnable = null
    emitRunnable = null
    workerHandler = null
    workerThread?.quitSafely()
    workerThread = null
    mainHandler = null
    eventSink = null
    sequence = 0
    sourceTick = 0
    windowSampleCount = 0
    windowSampleSum = 0.0
    windowDroppedCount = 0
    lastSampleValue = Double.NaN
  }

  private fun demoSampleValue(tick: Int): Double {
    val wave = sin(tick * 0.15) * 50.0
    val counter = (tick % 10).toDouble()
    return wave + counter
  }

  companion object {
    private const val SOURCE_RATE_HZ = 60
    private const val DELIVERED_RATE_HZ = 4
    private const val DELIVERY_WINDOW_MS = 250
    private const val SAMPLE_EPSILON = 0.01
  }
}
